src/eval.rs

```diff
@@ -75,48 +68,60 @@ fn eval(program: &Program, input: &str) -> Result<(bool, Env), RuntimeError> {
     // create initial state
     let mut env = init_env(program);
 
-    // TODO - figure out what to do with input ("consume each symbol???")
-    // consume each symbol
-
     // evaluate the final condition
+    for sym in input.chars() {
+        match &program.action.0 {
+            Some(id) => env.insert(id.clone(), Value::Sym(Symbol(sym))),
+            _ => continue,
+        };
+    }
     for stmt in &program.action.1 {
         eval_stmt(stmt, &mut env, &program)?;
     }
-    eval_expr(&program.accept, &env, &program)?;
+    let accept = eval_expr(&program.accept, &env, &program)?;
 
-    Ok((false, env))
+    match accept {
+        Value::Bool(b) => Ok((b, env)),
+        _ => Err(RuntimeError::TypeError),
+    }
 }
 
 fn cast(v: i64, range: Range<i64>, overflow: Overflow) -> Result<Value, RuntimeError> {
     let lower = range.start;
     let upper = range.end;
     let val = match overflow {
-        Overflow::Fail => {
-            if v > upper {
-                Err(RuntimeError::OutOfRange)
-            } else if v < lower {
-                Err(RuntimeError::OutOfRange)
-            } else {
-                Ok(Value::Num(v, range))
-            }
-        }
+        // Overflow::Fail => {
+        //     if v >= upper {
+        //         Err(RuntimeError::OutOfRange)
+        //     } else if v < lower {
+        //         Err(RuntimeError::OutOfRange)
+        //     } else {
+        //         Ok(Value::Num(v, range))
+        //     }
+        // }
         Overflow::Saturate => {
-            if v > upper {
+            if v >= upper {
                 Ok(Value::Num(upper - 1, range))
             } else if v < lower {
                 Ok(Value::Num(lower, range))
             } else {
                 Ok(Value::Num(v, range))
             }
         }
-        Overflow::Wraparound => Ok(Value::Num(
-            ((v - lower).abs() % (upper - lower) + lower),
+        // Overflow::Wraparound => Ok(Value::Num(
+        //     (v - lower).abs() % (upper - lower) + lower,
+        //     Range {
+        //         start: lower,
+        //         end: upper,
+        //     },
+        // )),
+        _ => Ok(Value::Num(
```

Comment:
restore the `Overflow::Wraparound` case. **You shouldn't merge Fail and Wraparound!**